File: graph/graph.ml
open List
open Printf

type 'a node = {
  value: 'a;
  mutable succ: 'a node list;
}
type 'a graph = { mutable roots: 'a node list}

module SSet = Set.Make(String)

let empty_graph () = {roots=([]: string node list)}

let find_node r value =
  let visited = ref SSet.empty in
  let rec find_x curr =
    if SSet.mem curr.value !visited then None
    else begin
      visited := SSet.add curr.value !visited;
      if curr.value = value then Some curr
      else begin
        let rec handle_lst lst =
          match lst with
          | [] -> None
          | x::lst ->
            match find_x x with
            | None -> handle_lst lst
            | Some x as res -> res
        in
          handle_lst curr.succ
      end
    end
  in
    find_x r

let rec find_lst_opt pred lst =
  match lst with
  | [] -> None
  | x::lst ->
    let res = pred x in
    match res with
    | Some x -> res
    | None -> find_lst_opt pred lst

let add_edge g a b =
  let nodeb = {value=b; succ=[]} in
  let res = find_lst_opt (fun x -> find_node x a) g.roots in
  match res with
  | None ->
    (*printf "add edge %s -> %s, empty\n" a b;*)
    let nodea = {value=a; succ=[nodeb]} in
    g.roots <- nodea :: g.roots;
  | Some x ->
    (*printf "add edge %s -> %s, found x = %s\n" a b x.value;*)
    x.succ <- nodeb :: x.succ

let g = empty_graph ()
let (->>) a b =
  add_edge g a b

let rpt_print n c =
  for i = 0 to n do
    printf "%c" c
  done

let print_graph g =
  iter (fun r ->
    printf "root: %s\n" r.value;
    let visited = ref SSet.empty in
    let rec print_node n width =
      printf "%s" n.value;
      if SSet.mem n.value !visited then printf "\n"
      else begin
        visited := SSet.add n.value !visited;
        let first = ref true in
        iter (fun succ ->
          if ! first then begin
            first := false;
            printf " -> ";
            print_node succ (width+5)
          end else begin
            rpt_print (width-1) ' ';
            printf "|\n";
            rpt_print (width-1) ' ';
            printf "+-> ";
            print_node succ (width+5)
          end
        ) n.succ;
        if List.length n.succ = 0 || List.length n.succ > 1 then
          printf ".\n"
      end
    in
      print_node r 0
  ) g.roots

let test () =
  "a" ->> "b";
  "b" ->> "c";
  "c" ->> "d";
  "d" ->> "a";
  "b" ->> "e";
  "e" ->> "f";
  "f" ->> "g";
  "g" ->> "e";
  "c" ->> "b";
  "g" ->> "h";
  "h" ->> "i";
  print_graph g

let _ =
  test ()

File: graph/graph.jl
module GraphMod

import Base.==

export Node, Graph, addedge, show, make_data

mutable struct Node
    pred::Set # = Set()
    succ::Set # = Set()
    name::String
    no::Int
end

function Node(; name::String)
    Node(Set(), Set(), name, 0)
end

mutable struct Graph
    roots::Set{Node}
end

function ==(x::Node, y::Node)
    x.name == y.name
end

function hash(x::Node)
    hash(x.name)
end

function findnode(g::Graph, x::Node)
    visited = Set{Node}()
    function inner_find(r::Node)
        if r in visited
            return nothing
        end
        union!(visited, [r])
        if r == x
            return r
        end
        for elt = r.succ
            res = inner_find(elt)
            if res != nothing
                return res
            end
        end
        return nothing
    end

    for r = g.roots
        res = inner_find(r)
        if res != nothing
            return res
        end
    end
    return nothing
end

function addedge(g::Graph, x::Node, y::Node)
    res = findnode(g, x)
    if res == nothing
        union!(g.roots, [x])
        res = x
    end
    resy = findnode(g, y)
    resy == nothing && (resy = y)
    union!(res.succ, [resy])
    union!(resy.pred, [res])
end

function addedge(g::Graph, x::String, y::String)
    addedge(g, Node(name=x), Node(name=y))
end

function show(io::IO, x::Node)
    visited = Set{Node}()
    function inner_show(n::Node, indent)
        println(io, ' '^indent, n.name)
        if n in visited
            return
        end
        union!(visited, [n])
        for elt = n.succ
            inner_show(elt, indent+1)
        end
    end

    inner_show(x, 0)
end

function show(io::IO, g::Graph)
    for r = g.roots
        println("=====")
        show(io, r)
    end
end

function main()
    g = Graph(Set())
    addedge(g, Node(name="a"), Node(name="b"))
    addedge(g, Node(name="a"), Node(name="c"))
    addedge(g, Node(name="b"), Node(name="c"))
    addedge(g, Node(name="c"), Node(name="a"))
    addedge(g, Node(name="c"), Node(name="d"))
    addedge(g, Node(name="d"), Node(name="a"))
    addedge(g, Node(name="b"), Node(name="e"))
    show(stdout, g)
end

#=
entry -- b1 -- b2 -- exit
         |               ^
         |           b5 -+
         +---- b3 -- b4 -- b6
                     ^     |
                     |     |
                     +-----+
=#
function make_data()
    g = Graph(Set())
    addedge(g, "entry", "b1")
    addedge(g, "b1", "b2")
    addedge(g, "b1", "b3")
    addedge(g, "b2", "exit")
    addedge(g, "b3", "b4")
    addedge(g, "b4", "b6")
    addedge(g, "b6", "b4")
    addedge(g, "b4", "b5")
    addedge(g, "b5", "exit")
    return g
end

# main()

end # end module Graph